RecoTracker/MkFit/plugins/MkFitOutputConverter.cc

```diff
@@ -0,0 +1,580 @@
+
+  LogTrace("MkFitOutputConverter") << "Number of candidates " << mkfitOutput.candidateTracks().size();
+
+  int candIndex = -1;
+  for (const auto& cand : candidates) {
+    ++candIndex;
+    LogTrace("MkFitOutputConverter") << "Candidate " << candIndex << " pT " << cand.pT() << " eta " << cand.momEta()
+                                     << " phi " << cand.momPhi() << " chi2 " << cand.chi2();
+
+    // hits
+    edm::OwnVector<TrackingRecHit> recHits;
+    const int nhits = cand.nTotalHits();  // what exactly is the difference between nTotalHits() and nFoundHits()?
+    bool lastHitInvalid = false;
+    for (int i = 0; i < nhits; ++i) {
+      const auto& hitOnTrack = cand.getHitOnTrack(i);
+      LogTrace("MkFitOutputConverter") << " hit on layer " << hitOnTrack.layer << " index " << hitOnTrack.index;
+      if (hitOnTrack.index < 0) {
+        // What is the exact meaning of -1, -2, -3?
+        // In order to use the regular InvalidTrackingRecHit I'd need
+        // a GeomDet (and "unfortunately" that is needed in
+        // TrackProducer).
+        //
+        // I guess we could take the track state and propagate it to
+        // each layer to find the actual module the track crosses, and
+        // check whether it is active or not to be able to mark
+        // inactive hits
+        const auto* detLayer = detLayers.at(hitOnTrack.layer);
+        if (detLayer == nullptr) {
+          throw cms::Exception("LogicError") << "DetLayer for layer index " << hitOnTrack.layer << " is null!";
+        }
+        // Actually it is necessary to leave dealing with invalid hits to the TrackProducer?
+        //recHits.push_back(new InvalidTrackingRecHitNoDet(detLayer->surface(), TrackingRecHit::missing)); // let's put them all as missing for now
+        lastHitInvalid = true;
+      } else {
+        recHits.push_back(indexLayers.getHitPtr(hitOnTrack.layer, hitOnTrack.index)->clone());
+        LogTrace("MkFitOutputConverter") << "  pos " << recHits.back().globalPosition().x() << " "
+                                         << recHits.back().globalPosition().y() << " "
+                                         << recHits.back().globalPosition().z() << " mag2 "
+                                         << recHits.back().globalPosition().mag2() << " detid "
+                                         << recHits.back().geographicalId().rawId() << " cluster "
+                                         << indexLayers.getClusterIndex(hitOnTrack.layer, hitOnTrack.index);
+        lastHitInvalid = false;
+      }
+    }
+
+    const auto lastHitId = recHits.back().geographicalId();
+
+    // MkFit hits are *not* in the order of propagation, sort by 3D radius for now (as we don't have loopers)
+    // TODO: Improve the sorting (extract keys? maybe even bubble sort would work well as the hits are almost in the correct order)
+    recHits.sort([](const auto& a, const auto& b) {
+      const auto aid = a.geographicalId();
```

Comment:
might be able to avoid these temporaries by combining with the next lines that define `asub` and `bsub`

Reply:
the compiler better take care of that

Reply:
I have seen cases before where it didn't.

Reply:
because of non trivial copy-constructor or destructor

Reply:
I anyway combined the lines.